File: app/src/main/java/openeyes/drawalive/seven/openeyes/Preview1.java
package openeyes.drawalive.seven.openeyes;

import android.content.Context;
import android.graphics.Bitmap;
import android.hardware.Camera;
import android.util.Log;

import java.util.List;

import openeyes.drawalive.seven.openeyes.filter.BitmapFilter;
import openeyes.drawalive.seven.openeyes.filter.Kernel3x3Filter;
import openeyes.drawalive.seven.openeyes.filter.SobelOperator;

import static android.content.Context.CAMERA_SERVICE;

// Ref:
// Camera      >> https://developer.android.com/training/camera/cameradirect.html#TaskOpenCamera
// CameraExternalIntent: https://developer.android.com/training/camera/photobasics.html
// N21->Bitmap >> https://stackoverflow.com/questions/35826709/yuv-nv21-image-converting-to-bitmap/35852318#35852318

// AutoWhiteBalance:
//   https://stackoverflow.com/questions/35439159/camera2-api-set-custom-white-balance-temperature-color
// YUV420888 to RGB:
//   http://www.polarxiong.com/archives/Android-YUV_420_888%E7%BC%96%E7%A0%81Image%E8%BD%AC%E6%8D%A2%E4%B8%BAI420%E5%92%8CNV21%E6%A0%BC%E5%BC%8Fbyte%E6%95%B0%E7%BB%84.html
// Set Orientation:
//   https://github.com/googlesamples/android-Camera2Basic/blob/master/Application/src/main/java/com/example/android/camera2basic/Camera2BasicFragment.java
class Preview1 extends Preview implements Camera.PreviewCallback {
   Preview1(Context context) {
      super(context);
   }

   @Override
   public boolean safeCameraOpen() {
      boolean qOpened = false;
      try {
         setCamera(Camera.open(Camera.CameraInfo.CAMERA_FACING_FRONT));
         qOpened = (mCamera != null);
      } catch (Exception e) {
         try {
            setCamera(Camera.open(Camera.CameraInfo.CAMERA_FACING_BACK));
            qOpened = (mCamera != null);
         } catch (Exception e2) {
            Log.e(TAG, "failed to open Camera");
            e.printStackTrace();
         }
      }
      return qOpened;
   }

   @Override
   public void stopPreview() {
      if (mCamera == null) return;
      mCamera.stopPreview();
   }

   @Override
   public void stopPreviewAndFreeCamera() {
      if (mCamera != null) {
         mCamera.stopPreview();
         mCamera.release();
         mCamera = null;
      }
   }

   protected Camera mCamera;
   private List<Camera.Size> mSupportedPreviewSizes;

   private void setCamera(Camera camera) {
      if (mCamera == camera) { return; }
      stopPreviewAndFreeCamera();

      mCamera = camera;

      if (mCamera != null) {
         List<Camera.Size> localSizes = mCamera.getParameters().getSupportedPreviewSizes();
         mSupportedPreviewSizes = localSizes;

         try {
            mCamera.setDisplayOrientation(0);
            mCamera.setPreviewCallback(this);
         } catch (Exception e) {
            e.printStackTrace();
         }
         mCamera.startPreview();
      }
   }

   @Override
   public void onPreviewFrame(byte[] bytes, Camera camera) {
      Camera.Parameters params = camera.getParameters();
      Camera.Size size = params.getPictureSize();
      int w = size.width, h = size.height;
      Bitmap bmp = yuvToBitmap(bytes, w, h, params.getPreviewFormat());
      bmp = rotatedBitmap(bmp);
      processFrame(bmp);
   }

   private void processFrame(Bitmap bmp) {
      BitmapFilter filter = null;
      filter = new SobelOperator();
      //filter = new Kernel3x3Filter(new double[]{-1,0,0, 0,1,0, 0,0,0});
      //filter = new Kernel3x3Filter(new double[]{-1,0,0, 0,0,0, 0,0,1});
      //filter = new HistogramEqualizationFilter();
      handler.post(new FilterAction(this, filter, bmp));
   }

   private class FilterAction implements Runnable {

      BitmapFilter filter;
      Bitmap bmp;
      Preview1 preview;

      public FilterAction(Preview1 _preview, BitmapFilter _filter, Bitmap _bmp) {
         preview = _preview;
         filter = _filter;
         bmp = _bmp;
      }

      @Override
      public void run() {
         if (filter == null) {
            preview.getImageView().setImageBitmap(bmp);
         } else {
            preview.getImageView().setImageBitmap(filter.act(bmp));
         }
      }
   }
}

File: app/src/main/kotlin/openeyes/drawalive/seven/openeyes/Preview1.kt
package openeyes.drawalive.seven.openeyes

import android.content.Context
import android.graphics.Bitmap
import android.hardware.Camera
import android.util.Log

import openeyes.drawalive.seven.openeyes.filter.BitmapFilter
import openeyes.drawalive.seven.openeyes.filter.Kernel3x3Filter
import openeyes.drawalive.seven.openeyes.filter.SobelOperator

internal class Preview1(context: Context) : Preview(context), Camera.PreviewCallback {

    protected var mCamera: Camera? = null
    private var mSupportedPreviewSizes: List<Camera.Size>? = null

    override fun safeCameraOpen(): Boolean {
        var qOpened = false
        try {
            setCamera(Camera.open(Camera.CameraInfo.CAMERA_FACING_FRONT))
            qOpened = mCamera != null
        } catch (e: Exception) {
            try {
                setCamera(Camera.open(Camera.CameraInfo.CAMERA_FACING_BACK))
                qOpened = mCamera != null
            } catch (e2: Exception) {
                Log.e(Preview.TAG, "failed to open Camera")
                e.printStackTrace()
            }

        }

        return qOpened
    }

    override fun stopPreview() {
        if (mCamera == null) return
        mCamera!!.stopPreview()
    }

    override fun stopPreviewAndFreeCamera() {
        if (mCamera != null) {
            mCamera!!.stopPreview()
            mCamera!!.release()
            mCamera = null
        }
    }

    private fun setCamera(camera: Camera) {
        if (mCamera === camera) {
            return
        }
        stopPreviewAndFreeCamera()

        mCamera = camera

        if (mCamera != null) {
            val localSizes = mCamera!!.parameters.supportedPreviewSizes
            mSupportedPreviewSizes = localSizes

            try {
                mCamera!!.setDisplayOrientation(0)
                mCamera!!.setPreviewCallback(this)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            mCamera!!.startPreview()
        }
    }

    override fun onPreviewFrame(bytes: ByteArray, camera: Camera) {
        val params = camera.parameters
        val size = params.pictureSize
        val w = size.width
        val h = size.height
        var bmp = yuvToBitmap(bytes, w, h, params.previewFormat)
        bmp = rotateBitmap(bmp!!)
        processFrame(bmp)
    }

    private fun processFrame(bmp: Bitmap) {
        var filter: BitmapFilter? = null
        filter = SobelOperator()
        //filter = new Kernel3x3Filter(new double[]{-1,0,0, 0,1,0, 0,0,0});
        //filter = new Kernel3x3Filter(new double[]{-1,0,0, 0,0,0, 0,0,1});
        //filter = new HistogramEqualizationFilter();
        handler.post(FilterAction(this, filter, bmp))
    }

    private inner class FilterAction(internal var preview: Preview1, internal var filter: BitmapFilter?, internal var bmp: Bitmap) : Runnable {

        override fun run() {
            if (filter == null) {
                preview.imageView.setImageBitmap(bmp)
            } else {
                preview.imageView.setImageBitmap(filter!!.act(bmp))
            }
        }
    }
}
